src/code/views/image-browser-view.coffee

```diff
@@ -0,0 +1,172 @@
+OpenClipart = require '../utils/open-clipart'
+tr = require '../utils/translate'
+
+{div, input, button, img, i, a} = React.DOM
+
+ImageSearchResult = React.createFactory React.createClass
+  displayName: 'ImageSearchResult'
+  getInitialState: ->
+    loaded: false
+  componentDidMount: ->
+    image = new Image()
+    image.src = @props.imageInfo.image
+    image.onload = =>
+      @setState loaded: true
+  clicked: ->
+    @props.clicked @props.imageInfo
+  render: ->
+    src = if @state.loaded then @props.imageInfo.image else 'img/bb-chrome/spin.svg'
+    (img {src: src, onClick: @clicked, title: @props.imageInfo.title})
+
+PreviewImage = React.createFactory React.createClass
+  displayName: 'ImageSearchResult'
+  cancel: (e) ->
+    e.preventDefault()
+    @props.addImage null
+  addImage: ->
+    @props.addImage @props.imageInfo
+  render: ->
+    (div {},
+      (div {className: 'image-browser-header'}, 'Preview Your Image')
+      (div {className: 'image-browser-preview-image'},
+        (img {src: @props.imageInfo.image})
+        (a {href: '#', onClick: @cancel},
+          (i {className: "fa fa-close"})
+          'cancel'
+        )
+      )
+      (div {className: 'image-browser-preview-add-image'},
+        (button {onClick: @addImage}, 'Add Image')
+      )
+      (div {style: {clear: 'both', marginTop: 10}}, 'TBD: Metadata')
+    )
+
+ImageSearch = React.createFactory React.createClass
+  displayName: 'ImageSearch'
+
+  getInitialState: ->
+    searching: false
+    searched: false
+    internalLibrary: @props.protoNodes
+    internalResults: @props.protoNodes
+    externalResults: []
+    selectedImage: null
+
+  changed: ->
+    @search limitResults: true, useTimeout: true
+  
+  showAllMatches: ->
+    @search limitResults: false, useTimeout: false
+    
+  search: (options) ->
+    query = $.trim @refs.search.getDOMNode().value
+    validQuery = query.length > 0
+    
+    queryRegEx = new RegExp query, 'i'
+    internalResults = _.filter @props.protoNodes, (node) ->
+      queryRegEx.test node.title
+
+    @setState
+      query: query
+      searchable: validQuery
+      searching: validQuery
+      searchingAll: validQuery and not options.limitResults
+      searched: false
+      internalResults: internalResults
+      externalResults: []
+      numExternalMatches: 0
+
+    clearTimeout @searchTimeout
+    search = =>
+      OpenClipart.search query, options, (results, numMatches) =>
+        @setState
+          searching: false
+          searched: true
+          externalResults: results
+          numExternalMatches: numMatches
+    if options.useTimeout
+      @searchTimeout = setTimeout search, 1000
+    else
+      search()
+
+  componentDidMount: ->
+    @refs.search.getDOMNode().focus()
+
+  imageClicked: (imageInfo) ->
+    @setState selectedImage: imageInfo
+
+  addImage: (imageInfo) ->
+    if imageInfo
+      @props.addToPalette imageInfo
+    @setState selectedImage: null
+    
+  render: ->
```

Comment:
This render method seems a bit long in the tooth.  Any thoughts for how to simplify?